### DISCUSSION TITLE
Flash loans

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#170] [General]
Adding support for flash loans would be cool, but I worry that we will not have bandwidth for this (implementing this won't be trivial).  I would regard this as a priority 2, or even a priority 3. What do you guys think?

If we were to implement this, I imagine the implementation would look something like this:

- We would have a per-token mapping that would map token addresses to boolean flags, such that we don't allow any token to be flash loaned, willy-nilly
- All tokens are non-flash-loanable, by default
- The implementation would follow a standard like [ERC-3156](https://eips.ethereum.org/EIPS/eip-3156)

**Comment by maxdesalle** [comment]
I'm not familiar enough with smart contract development to have an opinion here, but an interesting data point is how much liquidity there is in stablecoins in Sablier, as we should obviously be very cautious about allowing flash loans for other types of tokens.

According to Etherscan ([here are the contract addresses](https://docs.sablier.finance/protocol/guides/chains), for reference), there is about $4.62M in stablecoins (USDC, CUSDC, USDT and DAI) across all our deployments (Ethereum, Polygon...).

As our protocol grows, and we start to get more and more users, I imagine this amount could become much larger and reach the tens of millions of dollars, maybe more. If that's the case, flash loans might prove to be a very profitable venture for a potential future Sablier DAO.

Is there any way we could add this later on to the protocol @paulrberg? As a sort of "add-on"? Not familiar enough with the contracts to know if this would be possible.

**Comment by maxdesalle** [comment]
Maybe a stupid idea, but wouldn't it be interesting if Sablier users received the flash loan fees when their liquidity is used for a flash loan?

This would help us build a strong moat as users would be incentivized to stream stablecoins (they could even stream cUSDC to *compound* (pun intended) the rewards). I guess this must be hard to implement technically, though, and we would get less revenue for the DAO. The latter is OK for me, as we should always prioritize growth over profit (similar to how Amazon operates).

**Reply by PaulRBerg** [reply]
This is an interesting idea. @razgraf has also proposed this fee-splitting mechanism during our latest all-hands call.

The problem is that in practice this is impossible to implement in Solidity. Calculating each user's portion of a token in the smart contract pool has complexity O(n), which in less fancy terms means that you have to iteratively go through each stream to see how much is owned by whom (assuming that we had a reverse lookup functionality at the smart contract level, which we don't). Anything O(n) in Solidity is liable to overflow the block gas limit.

**Comment by andreivladbrg** [comment]
My position here is the same as it was when we first talked about this topic in the summer, i.e. is worth implementing this because it would be a honest profitable venture, giving it a priority 2. I think we can make stablecoins flash-loanable at start, regarding DAO tokens we have to reach out to them.

> Is there any way we could add this later on to the protocol @paulrberg? As a sort of "add-on"? Not familiar enough with the contracts to know if this would be possible.

Unfortunately no, as long we don't use [proxy upgradable pattern](https://docs.openzeppelin.com/upgrades-plugins/1.x/proxies), which we don't. 

> Maybe a stupid idea, but wouldn't it be interesting if Sablier users received the flash loan fees when their liquidity is used for a flash loan?

The idea sounds nice, but the complexity which would be added is not worth it.

**Reply by PaulRBerg** [reply]
Seconded. It would be nice to implement this, not least because of the fees (though I am not sure how significant they would be).

My only reservation is the implementation effort, which has to be balanced against the other issues currently open in the repo.

> Is there any way we could add this later on to the protocol @paulrberg? As a sort of "add-on"? 

Unfortunately, no, because our contracts won't be upgradeable, and there's a million and one reasons for which we want our contracts non-upgradeable.

**Reply by PaulRBerg** [reply]
As per a follow-on f2f discussion with @andreivladbrg, I have realized that the fees that could be obtained from flash loans are non-negligible:

Here are some back-of-the-napkin calculations based on conservative figures:

- Assume $1m USDC TVL in Sablier V2
- Assume 1 successful flash loan trade per day
- Assume 0.05% (almost twice as low as on Aave, which charges 0.09%)

This would mean:

- 500 USDC daily revenue
- ~15k USDC monthly revenue
- ~180k USDC yearly revenue

This is good money, so we should definitely consider adding support for flash loans in V2.

**Comment by PaulRBerg** [comment]
Caveat if we end up implementing flash loans.

We should update our [terms of service](https://sablier.finance/terms-of-service) and/ or our [protocol disclaimer](https://sablier.finance/protocol-disclaimer) to disclose to users that their are funds are used as liquidity for flash loans.

Interesting to note, however, than Aave doesn't mention flash loans in their [terms of use](https://aave.com/term-of-use/) (I think they definitely should).

**Comment by PaulRBerg** [comment]
Locking as this was implemented in #277.

For discussing the business implications of flash loans, we have discussion #278 still open.
